File: C++/intersection_of_two_linked_lists.cpp
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int l1 = 0;
        int l2 = 0;
        ListNode* cur1 = headA;
        ListNode* cur2 = headB;
        
        while (cur1 != nullptr) {
            l1++;
            cur1 = cur1->next;
        }
        while (cur2 != nullptr) {
            l2++;
            cur2 = cur2->next;
        }
        cur1 = headA;
        cur2 = headB;

        if (l1 > l2) {
            int dif = l1 - l2;
            while (dif > 0) {
                cur1 = cur1->next;
                dif--;
            }
        } else if (l1 < l2) {
            int dif = l2 - l1;
            while (dif > 0) {
                cur2 = cur2->next;
                dif--;
            }
        }
        while (cur1 != cur2) {
            cur1 = cur1->next;
            cur2 = cur2->next;
        }
        return cur1;
    }
};

File: PHP/intersection_of_two_linked_lists.php
<?php
class ListNode
{
    public $val = 0;
    public $next = null;
    function __constructor($val)
    {
        $this->val = $val;
    }
}

class Solution
{
    /**
     * @param ListNode $headA
     * @param ListNode $headB
     * @return ListNode
     */
    function getIntersectionNode($headA, $headB)
    {
        $l1 = 0;
        $l2 = 0;
        $first = $headA;
        $second = $headB;
        while ($first !== null) {
            $l1++;
            $first = $first->next;
        }
        while ($second !== null) {
            $l2++;
            $second = $second->next;
        }
        $first = $headA;
        $second = $headB;
        if ($l1 > $l2) {
            $dif = $l1 - $l2;
            while ($dif--) {
                $first = $first->next;
            }
        } else {
            $dif = $l2 - $l1;
            while ($dif--) {
                $second = $second->next;
            }
        }
        while ($first !== $second) {
            $first = $first->next;
            $second = $second->next;
        }
        return $first;
    }
}

// Выбор между операторами != и !== зависит от контекста и того, 
// что именно вы хотите проверить — значения или значения и тип данных.